xdsl/interpreters/snitch_stream.py

```diff
@@ -0,0 +1,169 @@
+from collections.abc import Sequence
+from dataclasses import dataclass
+from functools import reduce
+from itertools import accumulate
+from operator import mul
+from typing import Any
+
+from xdsl.dialects import snitch_stream
+from xdsl.interpreter import (
+    Interpreter,
+    InterpreterFunctions,
+    PythonValues,
+    ReturnedValues,
+    impl,
+    impl_terminator,
+    register_impls,
+)
+from xdsl.interpreters.riscv import RawPtr, RiscvFunctions
+from xdsl.interpreters.stream import (
+    ReadableStream,
+    WritableStream,
+)
+from xdsl.ir.affine import AffineExpr, AffineMap
+
+
+def indexing_map_from_bounds(bounds: Sequence[int]) -> AffineMap:
+    divs = tuple(accumulate(reversed(bounds), mul, initial=1))[-2::-1]
+    return AffineMap(
+        1,
+        0,
+        tuple(
+            AffineExpr.dimension(0).floor_div(div) % bound
+            if div != 1
+            else AffineExpr.dimension(0) % bound
+            for bound, div in zip(bounds, divs)
+        ),
+    )
+
+
+def offset_map_from_strides(strides: Sequence[int]) -> AffineMap:
+    if not strides:
+        # Return empty map to avoid reducing over an empty sequence
+        return AffineMap(1, 0, ())
+
+    return AffineMap(
+        len(strides),
+        0,
+        (
+            reduce(
+                lambda acc, m: acc + m,
+                (AffineExpr.dimension(i) * stride for i, stride in enumerate(strides)),
+            ),
+        ),
+    )
+
+
+@dataclass
+class StridePattern:
+    ub: list[int]
+    strides: list[int]
+
+    @property
+    def offset_expr(self) -> AffineExpr:
```

Comment:
Sorry, can you add some basic doc on this method, and the indexing and offset `map_from` methods above?

Maybe just adding comments to the test would work best? Not sure, but I feel there are a lot of abstraction levels "meeting" here.